### PR TITLE
Encode act_fn as hparam, use scratch buffers

### LINKED ISSUE #10: QuickGELU - not SOTA ?
I just read through the [open_clip readme](https://github.com/mlfoundations/open_clip#pretrained-model-interface) and found this section:
> NOTE: Many existing checkpoints use the QuickGELU activation from the original OpenAI models. This activation is actually less efficient than native torch.nn.GELU in recent versions of PyTorch. The model defaults are now nn.GELU, so one should use model definitions with -quickgelu postfix for the OpenCLIP pretrained weights. All OpenAI pretrained weights will always default to QuickGELU. One can also use the non -quickgelu model definitions with pretrained weights using QuickGELU but there will be an accuracy drop, for fine-tune that will likely vanish for longer runs.

does that mean, that we should not use QuickGELU/insert it as a hyperparameter ?

### PR DESCRIPTION
**Pull request by monatis**
Closes #10 and #12

Will be open for some time for regression tests.

### CODE REVIEW

File: clip.cpp [hunk 1]
```diff
@@ -883,7 +908,14 @@ bool clip_text_encode(
                        ggml_repeat(ctx0, model.layers[il].ff_i_b, cur),
                        cur);
 
-        cur = ggml_gelu_quick_inplace(ctx0, cur);
+        if (ctx->use_gelu)
+        {
+            cur = ggml_gelu_quick_inplace(ctx0, cur);
```
Comment (Green-Sky) [inline clip.cpp:913]: :)

**Review by Green-Sky — CHANGES_REQUESTED** [review]

File: clip.cpp (hunk 1 above)
Comment (monatis) [reply clip.cpp]: Quick GELU for everyone!11! :D

**Review by Green-Sky — APPROVED** [review]
gonna test this later :)
the reduces memory usage looks very good

**Comment by Green-Sky** [comment]
can you align the model hyperparameter output?
```
clip_model_load: num_positions   = 77
clip_model_load: t_hidden_size  = 512
clip_model_load: t_n_intermediate  = 2048
clip_model_load: t_n_head  = 8
clip_model_load: t_n_layer = 12
clip_model_load: image_size = 224
clip_model_load: patch_size   = 32
clip_model_load: v_hidden_size  = 768
clip_model_load: v_n_intermediate  = 3072
clip_model_load: v_n_head  = 12
clip_model_load: v_n_layer = 12
clip_model_load: use_gelu     = 1
clip_model_load: ftype     = 1
```
looks very funky as is

**Comment by monatis** [comment]
yeah it definitely needs some polishing. will have a look.

**Comment by Green-Sky** [comment]
with pr:
```
$ bin/zsl -m ../models/laion_clip-vit-h-14-laion2b-s32b-b79k/ggml-model-f16.bin.new --image ../tests/red_apple.jpg --text dog --text apple --text human --text photo --text drawing --text red --text blue
...
apple = 0.1614
red = 0.1451
photo = 0.1450
human = 0.1417
drawing = 0.1395
blue = 0.1353
dog = 0.1320
```

master:
```
$ bin/zsl -m ../models/laion_clip-vit-h-14-laion2b-s32b-b79k/ggml-model-f16.bin.old --image ../tests/red_apple.jpg --text dog --text apple --text human --text photo --text drawing --text red --text blue
...
apple = 0.1592
red = 0.1445
photo = 0.1433
drawing = 0.1421
human = 0.1420
blue = 0.1358
dog = 0.1332
```

yea, values are better. but I'm still surprised how little difference this makes.

**Comment by monatis** [comment]
> I'm still surprised how little difference this makes

It may even be worse in rare cases, but as you said single-instance tests are not so informatory. We still need a proper benchmarking. Maybe I should prioritize it soon.

After merging this PR, I can take on it. And, we can write a Medium post once benchmarking is done and the image search example is merged.

**Comment by Green-Sky** [comment]
hm, i kinda miss the '=' or a ':'
```
n_vocab            49408
num_positions      77
t_hidden_size      1024
t_n_intermediate   4096
t_projection_dim   1024
t_n_head           16
t_n_layer          24

clip_model_load: vision model hparams
image_size             224
patch_size         14
v_hidden_size      1280
v_n_intermediate   5120
v_projection_dim   1024
v_n_head           16
v_n_layer          32

use_gelu           1
ftype              3
```

> It may even be worse in rare cases, but as you said single-instance tests are not so informatory. We still need a proper benchmarking. Maybe I should prioritize it soon.

yea, will have to look at common datasets, and make sure the models were not trained on the test data.

> After merging this PR, I can take on it. And, we can write a Medium post once benchmarking is done and the image search example is merged.

haha, yea.

**Comment by monatis** [comment]
> hm, i kinda miss the '=' or a ':'


Do you prefer that one? I just thought of printing it out in some kind of 'tabular layout' based on the column width, i.e., string length.

**Comment by Green-Sky** [comment]
> > hm, i kinda miss the '=' or a ':'
> 
> Do you prefer that one? I just thought of printing it out in some kind of 'tabular layout' based on the column width, i.e., string length.

hm, its fine i suppose. (if you fix the `image_size` )
